### PR TITLE
Fix skipped pipelines model

### PR DESCRIPTION
**Pull request by xoxys**
Fixes: https://github.com/woodpecker-ci/woodpecker/issues/2901

### CODE REVIEW

**Comment by xoxys** [comment]
Don't know if this is a good way to fix it. Please let me know what you think.

**Comment by xoxys** [comment]
Correct all pipelines are saved to the DB at https://github.com/woodpecker-ci/woodpecker/pull/2923/files#diff-80265a1351d5d4936ff03f75e08517b6cfcca8015c02e730ebd90408983a0200R58 as required to populate all env vars to the step builder

In the case pipelines are empty, we can also remove them from the DB again. Writing to the DB and remove it again afterward feels wrong in general, as it creates unnecessary DB calls.

**Comment by anbraten** [comment]
> In the case pipelines are empty, we can also remove them from the DB again. Writing to the DB and remove it again afterward feels wrong in general, as it creates unnecessary DB calls.

Agree. Only other option would be to parse the pipeline, store it and set the pipeline env vars like id afterwards.

**Comment by xoxys** [comment]
Yep would need a bit more work but I think thats the better way.

**Review by lafriks — APPROVED** [review]

**Comment by xoxys** [comment]
After looking into it again I have not found a solution yet. The env var handling for pipeline workflows is tied together with the substitution logic and the workflow metadata and I was not able to decouple it to (re)call it after creating the pipeline. I would need help here.

**Comment by xoxys** [comment]
Ok I would say we just delete empty pipelines from the DB for now as a hotfix and I create a followup issue. What do you think?

**Comment by xoxys** [comment]
The PR deletes the pipeline from the DB again after checking if it was skipped. To avoid leftovers in the DB (e.g. by interrupted processes, errors during the deletion etc.) I have added a new state `StateCreated`. This state can be used to easily filter leftovers in the DB. If the pipeline was created and was checked to be not skipped, the status is updated to `StatusPending`. 

Personally I think I prefer the alternative approach from https://github.com/woodpecker-ci/woodpecker/pull/2928 but happy to discuss.

**Comment by xoxys** [comment]
As discussed in https://github.com/woodpecker-ci/woodpecker/pull/2928,as a hotfix this approach is preferred.

File: server/pipeline/create.go [hunk 1]
```diff
@@ -81,9 +81,17 @@ func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline
 
 	if len(pipelineItems) == 0 {
 		log.Debug().Str("repo", repo.FullName).Msg(ErrFiltered.Error())
+		if err := _store.DeletePipeline(pipeline); err != nil {
+			return nil, err
```
[the remark this answers is not in the document]
Comment (xoxys) [reply server/pipeline/create.go:85]: What do you mean? The error is returned already.

[the remark this answers is not in the document]
Comment (xoxys) [reply server/pipeline/create.go:85]: Nah, no worries :smile: at least we were able to clarify it.

[the remark this answers is not in the document]
Comment (xoxys) [reply server/pipeline/create.go]: Done

File: server/store/store.go [hunk 2]
```diff
@@ -93,6 +93,8 @@ type Store interface {
 	CreatePipeline(*model.Pipeline, ...*model.Step) error
 	// UpdatePipeline updates a pipeline.
 	UpdatePipeline(*model.Pipeline) error
+	// DeletePipeline updates a pipeline.
```
Comment (lafriks) [inline server/store/store.go:96]:
```suggestion
	// DeletePipeline deletes a pipeline.
```

**Review by lafriks — APPROVED** [review]
As quickfix this is ok with me before we find a better way to fix this

**Pull request merged**